### ISSUE TITLE
Saving data on webserver

### ISSUE THREAD

**Issue by maltelau** [djnavarro/jaysire#8]
First, thanks for this cool package making javascript experiments feel way more elegant to write. I just wrote a 100 line script in a couple of hours that generated exactly the experiment I need -- Yay!

I've a couple of suggestions for improvements. First one is about saving the data on a (non gae) server. I'm from a European university with a data processing agreement in place with a major cloud provider, so I'd like the data to go directly there and not clutter the participant's pc and not touch google servers. In the jspsych documentation, they suggest sending it via POST request to a little php script that saves the data ([link](https://www.jspsych.org/overview/data/#storing-data-permanently-as-a-file)). Why not include that functionality in jaysire?

Would involve:
- **A php script to receive and save the data**

```php
<?php
// take the uploaded data
$post_data = json_decode(file_get_contents('php://input'), true);
// Write the contents to a new file
file_put_contents('/var/www/server_data/data_'.date('Y-m-d').'_'.uniqid().'csv', $post_data['filedata']);
?>
```

- **A fn_save_serverdata function to return the javascript needed to send the data**
```R
#' Return a javascript function to save data via a script on the webserver
#'
#' @export
#' @return Return a javascript function to save data via a script on the webserver
#'
#' @details The purpose of the \code{fn_save_serverscript()} is to return a
#' javascript function that, when called from within the jsPsych experiment,
#' will write the data to the server. This assumes that the experiment will
#' be run on a (php)script-enabled webserver. This way, you know the data will
#' never touch any other computer than the server you've presumably secured
#' and have data processing agreements in place for.
fn_save_serverscript <- function() {
  js_code(
    "function() {
       var data = {filedata: jsPsych.data.get().csv()};
       fetch('write_data.php', {
         method: 'POST',
         body: JSON.stringify(data),
         headers: new Headers({
           'Content-Type': 'application/json'})});
    }")
}
```
- **Writing the php script to the experiment folder if needed**
(just like the gae files)

- **A php-enabled webserver** (print a message to remind user?)
- **A folder on the server where the webserver has permissions to write**
Not sure about this, maybe print out a list of commands for the user to run? ie for apache
```bash
sudo mkdir /var/www/server_data
sudo chgrp www-data /var/www/server_data
sudo chmod 775 /var/www/server_data
```
- **(optionally) a way to upload the experiment files**
Could either be a reminder to upload the experiment/ folder, or a script to do it, ie with Rcurl::ftpUpload()

If you think this is a good idea, I can try to write it out and send a pull request

**Comment by djnavarro** [comment]
Very good idea. I'd welcome a PR if you'd like to write it but otherwise yes, I've added this to my to-do list (it's the start of teaching at the moment so it may take me a while)

**Comment by djnavarro** [comment]
closed by 35bc8a3a06997c161bd7b98f7047709e6b1ef90e
